mlx-nn/src/pooling.rs

```diff
@@ -0,0 +1,338 @@
+use std::{
+    iter::{once, zip},
+    sync::Arc,
+};
+
+use mlx_macros::ModuleParameters;
+use mlx_rs::{error::Exception, module::Module, ops::as_strided, Array};
+
+use crate::utils::SingleOrPair;
+
+/// Type alias for a pooling operation.
+type PoolingOp = dyn Fn(&Array, &[i32]) -> Result<Array, Exception>;
+
+/// Abstract pooling layer.
+///
+/// See also:
+///
+/// - [`MaxPool1d`]
+/// - [`MaxPool2d`]
+/// - [`AvgPool1d`]
+/// - [`AvgPool2d`]
+#[derive(Clone, ModuleParameters)]
+pub struct Pool {
+    /// Size of the pooling window
+    kernel_size: Vec<i32>,
+
+    /// Stride of the pooling window
+    stride: Vec<usize>,
+
+    /// Axes to pool over
+    axes: Vec<i32>,
+
+    /// Pooling operation
+    ///
+    /// TODO: We have Arc here just to make it `Clone` and `Send`. Is this necessary?
+    pooling_op: Arc<PoolingOp>,
```

Comment:
One thing I'm not sure here is whether we should use `Arc`. My thinking right now is that this is going to be a trait object anyway and might as well just make it `Clone` and `Send`